### ISSUE TITLE
 [Feature request] Track Navigation support

### ISSUE THREAD

**Issue by megabugman** [organicmaps/organicmaps#1360]
After importing a track in KML file format, I miss the option to follow (navigate) along that path/track.
This feature is implemented in OsmAnd.

**Comment by biodranik** [comment]
This "follow" mode at least should be designed somehow in a safe way. Questions:

1. Is it for foot or for car navigation?
2. Voice? Directions? Vibrate? How it should work?
3. What to do if there is no map under the track?
4. Is the voice "you're too far from the track" notification enough?

**Comment by megabugman** [comment]
Hi, 
1. I like to use it for motorcycle but from my point of view car, bike, foot should be supported.
2. It should work as the normal routing. In normal routing a start and end point and stops (optionally) are included manually. A track normally includes a start/end point and may stops. From my understanding when deviating the application should use voice indication to the next/nearest point in the track. 
3. the same procedure as with "normal routing" what happens there.
4. not sure. The application should route to the nearest point until navigation is turned off.
thanks

**Comment by biodranik** [comment]
> 3. the same procedure as with "normal routing" what happens there.

There is no such situation in "normal routing". Because "normal routing" always uses underlying maps data for tracks. And it always knows crossings and directions.

If there is NO maps data under your track, nothing can be done with directions. If underlying maps data is incorrect, then you also won't get proper directions (example: there is a road under the track, but there is no crossing road on the map).

**Comment by biodranik** [comment]
See also #2384

**Comment by dandud100** [comment]
> This "follow" mode at least should be designed somehow in a safe way. Questions:
> 
> 1. Is it for foot or for car navigation?
> 2. Voice? Directions? Vibrate? How it should work?
> 3. What to do if there is no map under the track?
> 4. Is the voice "you're too far from the track" notification enough?

Maybe get a look at this[this](https://routes-navigation.de/) android app. It detects how far the starting point of the .gpx file first, if it is far away then it will tell that you should route with gmaps for example first as it is not a routing App for normal routes. When you are at the starting point it will give you voice instructions when following the track.

**Comment by biodranik** [comment]
It would be great to see some video + voice recordings on how other apps are doing it for foot and car/bicycle nav.

**Comment by Mike-0815** [comment]
Hello Alexander and the other devs,

I think, that principally shouldn't be very complicated to implement: 

After importing a (gpx/kml) track, OM should try to match it with underlying OSM street-data. If there is no underlying OSM street-data matchable (for example deep in the woods), OM shouldn't complain about that, and use and display the trackpoint(s) recorded in the track, on the map anyway.

If there -IS- underlying street-data available in an imported track, OM should (principally) navigate the user exactly in the same way, how navigation works in OM today, but along the track. 

If there is -NO- underlying street-data available, OM should navigate the user from any significant trackpoint to the next significant trackpoint (while always showing the map and the actual location-triangle) anyway. Significant means, that there will be a significant change in direccion between the last and the following trackpoint. OM should tell the user by a compass-like arrow on screen and voice direction, where to go. Voice direccions in the form of "go straight ahead", "turn slightly right", "turn right", "turn back" and so on.
Witch such given voice-direccions, anybody is able to navigate even with display turned off, easily, especially when using a Bluetooth-headset.

There's also an App called "Routes", witch is specialized in loading tracks/routes and guides the user along them. Maybe you'll take a look at it, when this the implementation of this feature become actual/concret.

Kind regards,
Mike

**Comment by d-damien** [comment]
An idea, not sure it's right : when too far from the track, use pathfinding to add a temporary track to the closest point of the track − remove the temp track when close enough.

But is it not possible to copy what is (was ?) being done in OsmAnd ?

**Comment by biodranik** [comment]
@cyber-toad are there any specific GPX-only elements that may be important for navigating over importing track? Does KML support something in this regard? It would be great to check what's used/popular on the market now, and make a list or table for GPX and KML of what we are missing now and need to support during the import/conversion.

**Comment by cyber-toad** [comment]
> @cyber-toad are there any specific GPX-only elements that may be important for navigating over importing track? Does KML support something in this regard? It would be great to check what's used/popular on the market now, and make a list or table for GPX and KML of what we are missing now and need to support during the import/conversion.

I'm not sure what elements for navigation do you need? I think we just need to build a line and track how close user is to this line and we have all geometry already?

**Comment by walterjwhite** [comment]
I would like to see this feature in OM.

For me, I would expect it to work the same with on or off road, if there is a trail or road to follow that and then limit the voice guidance to only when there is an intersection or change to another road.

That said, that may be easier to implement for on-road tracks than off-road where the road or trail may disappear.  If that happens, what should OM do, follow the track up to some pre-configured distance and beyond that, warn the user that the imported track is not following a known trail or road and manually stitch in that GPX segment?

Additionally, should there be an option to try to use nearby trails or roads prior to importing?  Perhaps if users don't want guidance, there could be another option which disables guidance and just shows the route on the map for the user to figure out.

Additionally, the original track is just a vehicle to transfer information into the app.  That said, perhaps there could be one last option which allows the user to save the track as an OM route to pull it up next time like a shortcut so that OM doesn't need to do the whole reimport process again?  This would be akin to the option in Google Maps to save a shortcut to the home screen.  It need not be the home screen, but within the app itself.

For routes that are prepared externally, this would allow OM users to use a track as if it were drawn within the app and benefit from the voice guidance and turn-by-turn directions.

In summary:

1. add option to 'snap' to underlying roads and trails within a given tolerance
2. add option to use segments of GPX track verbatim if there is no road or trail underneath (within the above tolerance) and for a given distance before failing or warning the user
3. add option to enable turn-by-turn navigation for imported GPX tracks
4. add option to save processed GPX track as a planned OM route for future navigation and reduce the need to reprocess the same GPX

**Comment by biodranik** [comment]
OM's philosophy is "let's keep it simple to use". Fewer options, more convenient, expected behavior.

**Comment by Mike-0815** [comment]
Exactly Alexander ! That's one of the "magic", the advantage over 99,9% of the navigation-apps OM has and it's philosophy should be kept strictly like this - easy2use, even at the cost of not implementing a function, but *NEVER* non-intuitive to use and many functions will make this much more difficult to keep. Really good software does even not need a manual !

**Comment by oleg-PK** [comment]
From [ platform/channel ]

> Hi, Alexander, and thank you for the lightning fast reply!
> 
> The motorcycle group that I'm in leads a bunch of group rides throughout the summer. We really want participants to have the route in case they get separated from the main group, or if we have to split into smaller groups.
> 
> The problem is that some of the participants are not particularly tech savvy. So, I'm trying to find an easy, inexpensive, app. to recommend for them to use.
> 
> I personally use OSMAnd as my GPS. It's an excellent app, but not the easiest to learn. It follows tracks extremely well, and like Organic Maps, works off line.
> 
> Tom Tom GO Ride does a nice job with following tracks, but needs a cell connection to work - something we don't always have. 
> 
> That's where we're at, and what I'm trying to do. 
> 
> Thanks again

**Comment by Scoo27** [comment]
hi
is there a date yet, when you will implement Track navigation?

**Comment by 0709wiwiwi** [comment]
With track navigation, you have these options.

- You follow the track on the map with TBT instructions based on the track geometry.
- You follow the track on the map with TBT instructions based on the (osm) street pattern.

1. You follow an ordinary track without additional editing. 
You either get too few instructions or a plethora of meaningless instructions at every track direction change. 

2. The track is snapped onto the road network via a router which adds the TBT instructions. 
Depending on the router engine, the export track will differ little or sometimes quite a lot from the original. 

To test the method 2.
Upload a quality (garbage in = garbage out) (record) track into the [RouteYou planner](https://www.routeyou.com/route/add)
The export gpx track fairly faithfully follows the original path even if no nearby osm path is known.
For export, select as Locus gpx. This gpx track is thus enriched with the TBT instructions.

After import into the Cruiser or Locus app, you immediately get a decent TBT gpx track navigation.
As long as you don't then let the app routers recalculate , you 100% faithfully follow the design. 
Other apps import a navtrack as plain gpx track without offering the extra.

More [Info](https://github.com/organicmaps/organicmaps/issues/5595#issuecomment-2128825902)

Translated with DeepL.com (free version)

**Comment by Scoo27** [comment]
hi @0709wiwiwi 
sorry, don't understand that.
I just want to navigate along an importet track with / in organicMaps.
What do you mean saying "after import into cruiser or Locus App" ?
(we can write in German if you want, via PN if this is possible on github)

**Comment by 0709wiwiwi** [comment]
> It would be great to see some video + voice recordings on how other apps are doing it for foot and car/bicycle nav.

Unless one like a longest day video, it seems more optimal to name apps that already (partly) offer this. 
Some apps are just briefly explored for a test.

1. Komoot app through the associated online web router offers mode 2. Komoot offers gpx track file export matched with the (osm) network. Off grid track sections are retained. The export track file does not contain TBT info.

2. RouteYou app through the associated online web router offers mode 2. RouteYou offers gpx track file export matched with the (osm) network. Offgrid tracks are preserved. The export (Locus gpx) navigation track contains TBT instructions. This format is directly usable after import for both Locus and Cruiser 100% faithful navigation.

3. Cruiser app. For simple track import offers mode 3. You have to manually check and update the result. The export gpx navigation track contains TBT instructions.

4. Locus app. After track import, offers mode 1 and 3. You need to manually check and update the result. 

5. OsmAnd app through the associated online web router. After track import, offers mode 1 and 2. I have not tested this app in depth. The app set-up is too complex for quick tests. The OsmAnd web gpx track export does not include TBT info. 

> is there a date yet, when you will implement Track navigation ?

Through Organic maps, neither option 1 nor option 2 is yet offered. The RouteYou web gpx navigation track export may eventually be a usefull external source before an in-house solution is ever potentially offered. Automatically faithfully converting a track into a TBT navigation track is delicate and quite complex.

**Comment by Scoo27** [comment]
Thank you for your feedback. I would like to use the navigation via carplay. I didn't mention this because Organic Maps supports it and I would like to use it primarily. As you say, TBT is unfortunately not supported (yet)
komoot is really very good, I already use it. Unfortunately, it doesn't support carplay at all, whereas with osmand you first have to pay to activate the function (as with many other providers). I am not yet familiar with the other apps mentioned and will see what is possible with iOS.



Translated with DeepL.com (free version)

**Comment by 0709wiwiwi** [comment]
The other apps:  Most Android and no carplay support.  - Locus recently have started to support IOS (beta).

**Comment by 0709wiwiwi** [comment]
As I noticed some new activity here (changed the title @rtsisyk ).
Find more info [gpx navigation track](https://github.com/devemux86/cruiser/discussions/250)
This reliable mode results in a 100% faithfull TBT track navigation  experience.
Attached into a gpz file additionally allows the transfer and display of any waypoint icon.

**Comment by Quantum-Future** [comment]
> This "follow" mode at least should be designed somehow in a safe way. Questions:
> 
>     1. Is it for foot or for car navigation?
> 
>     2. Voice? Directions? Vibrate? How it should work?
> 
>     3. What to do if there is no map under the track?
> 
>     4. Is the voice "you're too far from the track" notification enough?

1- It should work for foot, bicycle or car.

2- Just the usual way it works today when you are following a route you created.

3- Display a message offering to download the corresponding map for that zone.
     It's the same that happens when you scroll the map to another country or area and you don't have the map for that area.

4- Maybe yes, it's Ok for a first version.
    Later it could be added the possibility to calculate a temporary route the the nearest point of the route.

**Comment by HyGy** [comment]
What if we import a track we just select the closest point and navigate trough the gpx (and need an option to reverse the navigation direction). I think it does not require to plan a new route, just show the position on the map, and follow the track GPX track. If we press reverse then just navigate to the other way. The user can decide where to go when checking the screen with the real position.
I think do not need to select foot/car/motorcyle/bycicle. Couse I think it does not matter.

**Comment by Atemu** [comment]
Could an imported track perhaps be used to "bias" the regular routing algorithm in some way?

My use-case is exporting a route from brouter on the PC (because the routing for bicycles is unfortunately quite lacking/not configurable enough) and then have OM navigate me along that route on the phone.

Usually, OM actually gets quite close but there are a handful of key points along the route where it'd have to make a different decision. If it could take the imported track to influence the decision at those points, it might just work out well enough.

**Comment by 0709wiwiwi** [comment]
@Atemu 
So if I understand you correctly you just want to follow this Brouter track design ?
The Brouter exports the routing result also as a gpx navigation track. 
This is a gpx track which is enriched with all the TBT in it already present. 
A “compatible” receiver app does not need to recalculate to regenerate the TBT.
All too often a recalculation changes the original design slightly or sometimes even completely
More info about the gpx navigation track mode also exported by the BRouter can be found here.
https://github.com/devemux86/cruiser/discussions/250

**Comment by Atemu** [comment]
If it is possible to turn OM into such an "receiver" here, I think that could work.

Though sometimes I might in practice choose a slightly different route, take a detour etc. and not follow the route exactly. In those cases, I'd want my maps application to recalculate the route for me to get me back on track.

I've tried to nudge the OM routing algorithm to follow brouter's route using waypoints before but OM would often not recognise that I had passed a waypoint and try to navigate me backwards towards it. I'd like to avoid such situations.

**Comment by 0709wiwiwi** [comment]
@Atemu 
What you describe is possible in the mentioned app Autorecalculation can be either strict or non strict toward each next target via or shaping point. The recalculation uses the built-in offline BRouter. If you deviated from the traject frequent u-turn instructions to rejoin the design are avoided when possible.

**Comment by ilmc888** [comment]
After playing around with importing (and saving) gpx / tracks as bookmarks:

- It would be nice if you could actually select and highlight tracks as objects on the map (or when loading from bookmarks). Loading a track from bookmarks it seems it is represented as a single point (likely the centroid?). No distance is displayed either (in the bookmarks there is a length but this was not correct). 

- Following a track visually is fine for me, no TBT instructions needed (for now) but I still would like to start 'navigating' so that GPS speed, speed limits etc. are shown for the road i am on (except for ETA, remaining distance). Of course, it is mostly user convenience, as i could just navigate to a random point to achieve the same.

- After planning a route with way points it is still strange you loose the possibility of navigation just by saving and loading it in moments time (as i needed to quickly check something else). Recalculating the navigation based on the way points only would be fine for me (optionally we can save some settings along with the gpx file while planning, such as avoiding highways).

**Comment by Plugrol** [comment]
Any updates on this?

Especially for motorcycle routes this is a neat feature, also because there is no option to avoid "Kraftfahrstraßen" which are basically Highways with intersections in germany.

**Comment by Silverlan** [comment]
I would also like to see this implemented. I use brouter to plan my rides and OrganicMaps is my favorite app for navigation, but the lack of being able to navigate my pre-planned routes is unfortunately a deal-breaker for me and I have to resort to other apps.

**Comment by biodranik** [comment]
Is exporting GPX with a detailed track the only option supported by brouter? Does it allow passing route points using other ways?

**Comment by Silverlan** [comment]
> Is exporting GPX with a detailed track the only option supported by brouter? Does it allow passing route points using other ways?

The available export options in brouter are: GPX, KML, GeoJSON, CSV, FIT. You can optionally include the user-defined waypoints with the export:

<img width="552" height="390" alt="Image" src="https://github.com/user-attachments/assets/25e39109-c73d-45b8-b962-287009b89cfb" />

Here's a random example route:
https://bikerouter.de/#map=12/49.2573/3.1876/standard&lonlats=3.109655,49.258242;3.126646,49.29848;3.149296,49.306999;3.191868,49.318069

You can see the export options by clicking "Download" in the top navigation. I've attached the example route as GPX with waypoints:
[Villers-Cotterêts - Saint-Pierre-Aigle - 17 km, elev. gain 152 m.gpx.zip](https://github.com/user-attachments/files/21700686/Villers-Cotterets.-.Saint-Pierre-Aigle.-.17.km.elev.gain.152.m.gpx.zip)

**Comment by 0709wiwiwi** [comment]
The same gpx navigation track inclusive a short link to Essbee Brouter web.

[Villers-Cotterts - Saint-Pierre-Aigle - 17km.gpx.txt](https://github.com/user-attachments/files/21768365/Villers-Cotterts.-.Saint-Pierre-Aigle.-.17km.gpx.txt)

- Locus map shows you this clickable web link in the track properties.
  Short link services do not contain any prohibited characters such as `&`.

**Comment by ontheair81** [comment]
I planned a motorcycle tour. This was done directly within OM (not importing anyting). I saved the result as a track in OM.
Then I failed to start navigating this track, searched the web for a solution and found this thread.

If I understand everything correctly, this means:
- Planning can be done in OM.
- The planned route can be saved in OM.
- But navigating this pre-planned track (voice directions) is not possible. It has to be exported to another navigation app.

If I misunderstood something, please correct me! I really would love to use OM for our tours, but this seems not to be possible (with pre-planned tracks)?

**Comment by Quantum-Future** [comment]
Yes @ontheair81, it's correct as you said.
It seems weird/strange but it's because not all information needed to navigate the track is saved.
I'm not an expert on this, but as far as I know, there are several file formats and some technical difficulties, but the development team is working on that (I'm not part of it).

**Comment by 0709wiwiwi** [comment]
@ontheair81 @Quantum-Future 

A simple TBT gpx navigation track clearly demonstrates that this is perfectly feasible.
However, convincing app developers to implement this is **exceptionally difficult**. 

History. 

2014 Locus map. - Support requested. Possibly someday.
2016 BRouter. - Support requested. Possibly someday.
2020 Locus beta version. -  Import support. 
2020 Kurviger web. - Support requested.  "--" >>>> go to Cruiser gps app
2021 Locus official release. - Import support.  Export no . (2026)
https://forum.locusmap.eu/index.php?msg=49519
2022 Cruiser gps. - **Excellent** in & out support by improved Cruiser TBT gpx navigation track. 
https://github.com/devemux86/cruiser/discussions/250
2023 BRouter app/web. - Locus TBT gpx navigation track export. (Locus-style)
https://github.com/abrensch/brouter/pull/510#issue-1600100108
2025 BMaps Navigation.  Share as TBT gpx navigation track.
https://bmaps.nl/
2026 The Bikerouter planner TBT gpx navigation track export. (Locus-style)
2026 Routemaster GPS trackeditor. Does read TBT gpx navigation tracks. Convert in Garmin tcx & fit courses.
https://www.routemaster.app/software/routemaster.html
2026 Pitstopper.net routeplanner offers the export of TBT gpx navigation tracks.
https://pitstopper.net/blog/route-planner-beta.html
2026 Pitstopper.net converts (recorded) simple tracks into TBT gpx navigation tracks.
https://pitstopper.net/blog/match-your-ride-to-the-roads.html

**Comment by ontheair81** [comment]
> 2024-2025. - More apps ignore this simple mode and so continue to struggle.

I guess, the majority of developers are car drivers and they don't understand that motorbikers have different needs and expectations. Anyway, thank you @0709wiwiwi and @Quantum-Future for the explanation!

I will keep an eye on this thread, if this functionality will be added in the future. Would be really nice!

**Comment by AmiiiiKo** [comment]
As someone using a bicycle and electric scooter I have to take the safe route over the short route.
I think they underestimate the value of being able to follow a pre-planned track.

When I plan a route I utilise an existing map which highlights existing cycle paths and safety issues. I create a route on the map and export it... and I can't use it with organic maps without creating individual waypoints along that route. With other maps I can just click and go.

Organic maps has a superior navigation display and I would prefer to use it and request this feature be implemented.

**Comment by Atemu** [comment]
I think we have enough "me too"s.

**Comment by Semuria** [comment]
> ooh ooh I did read [@Semuria](https://github.com/Semuria) linked and now deleted comment anyway ;-) It turns out to be “there” too, just biting its own tail and going around in circles.
> 
> * Better just go and execute.  Simple reliable and robust result.
>   Apply as [gpx TBT navigation track](https://www.dropbox.com/scl/fi/niav72lr3td153u66axm3/Locus-3.50.1-Navigation-Commands-Icons_2p.doc?rlkey=l2zx9lzt1uw0k39snfdsub0fp&dl=0).
> 
> **Design a track.** Brouter web does export a track with built-in turn instructions as a gpx navigation track, thus providing 100% reliable navigation results. You just need to know how to select this in Brouter web. How? You can read about it here. [devemux86/cruiser#250 (comment)](https://github.com/devemux86/cruiser/discussions/250#discussioncomment-6508754) Unfortunately, the popular bikerouter web variant does NOT generate the required Locus new style mode yet!
> 
> **Import a simple track**. If you import a regular track, the OsmAnd attach to road tool works very well. Not entirely flawless, but definitely usable. However, this fine tool can only be usefully used in the OsmAnd app. Export as simple gpx navigation track suited for external apps is NOT offered. [osmandapp/OsmAnd#21936](https://github.com/osmandapp/OsmAnd/discussions/21936)

Rn I use komoot for this. There are popular presaved bike tracks already onboard and the possibility to create a track yourself and safe as gpx. for free in my region.

**Comment by 0709wiwiwi** [comment]
Import a simple track. 
If you import a regular track, the Komoot attach to road feature works definitely  very well. But the TBT is only offered in the Komoot app. Export as simple gpx navigation track suited for external apps is NOT offered as well.

**Comment by 0709wiwiwi** [comment]
To test the map matching or attach to road performance try next.

OsmAnd.
On sofa’ web test: Go to [osmand.net](https://www.osmand.net/map/#9/51.0000/5.0000)
Select Map > Left panel > Tap arrow icon (5) > Set route profile > Tap ‘approximate gpx’ > Select a decent quality track(record) > Tap ‘more information’ > Tap ‘turns’ > Browse the turns list > Find the turn positions on the map.

Bmaps navigation (bmaps.nl) is another new app.
Bmaps (offline) offers Valhalla map matching.

**Comment by Petak73** [comment]
Hi,
Attaching a GPX track to the roads and navigation it seems a rather difficult functionality to implement.
I have a much simpler method, which in some ways is even more convenient.
I import a GPX track in OrganicMaps, select a point on the track as my next stop (daily destination, petrol station etc.) and make OM create a route to that point and navigate. Obviously the route doen't always follow the GPX track but as soon as it deviates I simply follow the track and and the route is being automatically recalculated accordingly. The problem with this is that OM currently displays a very tiny GPX track line that is hardly visible at navigation mode/zoom levels.
Please see [this ](https://github.com/orgs/organicmaps/discussions/11535) post for more info.

**Comment by Petak73** [comment]
> > 2024-2025. - More apps ignore this simple mode and so continue to struggle.
> 
> I guess, the majority of developers are car drivers and they don't understand that motorbikers have different needs and expectations. Anyway, thank you [@0709wiwiwi](https://github.com/0709wiwiwi) and [@Quantum-Future](https://github.com/Quantum-Future) for the explanation!
> 
> I will keep an eye on this thread, if this functionality will be added in the future. Would be really nice!

I'm a motorcyclist myself and use pre-planned routes extensively and have tried a number of apps. Until few years ago the best implementation of planning a route at home on my laptop and then being navigated along it was by TomTom. They had a pretty good web route planner and the created routes were automatically synced to both dedicated TomTom devices and their nav app. Then you could simply open the route on the device/app and get a fully featured navigation - TBT, Voice, distances, etc. Unfortunately, their web interface became unusable after a massive "upgrade" a few years ago. They didn't fix it for quite some time and I lost interest.

The second best I've tried is OSMand. You can import a GPX track, ask the app to "attach it to the underlying roads" and then follow it. I works pretty well.

You can use Organic maps as well and i do but in a different way. You can import a GPX track and have it displayed on the map. Then you chose a point on the track - for exemple your daily destination or a petrol station and ask the app to navigate to this point. Then, whenever the route created by the app deviates from the GPX track you simply follow the track and OM recalculates the route accordingly. Obviously you can also select multiple points on the track and add them as "stops" along the route and then the calculated route will more or less follow the track. The problem is that the GPX tracks are displayed in OM with a very thin line which is barely visible on a motorcycle. If they fix that, the above method will work pretty wel for me. In fact, I would prefer to have the original GPX track visible at all times, even if I deviate slightly from it for whatever reason - error on the track, new streets, roundabouts, petrol stations, etc.

**Comment by Maranobarbaro** [comment]
I found Yesterday this app. I think the developer team did a great job but if is not possibile to follow a predefined saved track It limit a lot the usability... That feature Is not only a nice-to-have feature but It is more like a must-have feature for Who want to start from home to reach a predefined tour not having in mind the stress to define online the Path once arrived on the starting point. Sorry for my 2 cents

**Comment by Tablel352** [comment]
There are many pre-defined hiking tracks available online (for example in GPX format) and it's great they can be imported in Organic Maps. The feature to be able to use turn-by-turn navigation on those tracks would make this app a no-brainer (combined with the open source nature, focus on privacy, the use of OpenStreetMap and the user-friendlyness).

This screenshot made me already think the feature is available in Organic Maps: https://organicmaps.app/images/screenshots/dark.jpg. Unfortunately in the current version this only works for navitation to a certain end-point.

**Comment by Talkless** [comment]
> but if is not possibile to follow a predefined saved track It limit a lot the usability..

And not necessarily whole track as is, just stop points and final destination points could be enough too.

Intermediate tracks can be recalculated based on changed mapping, but for "scenic routes" the most important are (IMO) "key stop points" to go trough, and then let router decide the reset.

**Comment by 0709wiwiwi** [comment]
You shouldn't even settle for less. As It is already possible without further ado.  Strict or flexible ? As you like. 
Any apps can freely add and apply this supersimple mode ...anyone inclusive Organic.

[Gpx tbt navigation track.pdf](https://github.com/user-attachments/files/28702913/Gpx.tbt.navigation.track.pdf)

.

**Comment by JP7845** [comment]
@biodranik Hello, if you have a moment take a look at my posts below including this one, and let me know if this is of interest to the dev team. Thank you.


Maybe this is redundant, but I thought I would contribute another viewpoint...

### **Feature Concept: Track → Route Conversion with Adjustable Proximity**
**Core Idea**
Organic Maps should allow users **to convert any line (track)** — whether imported or recorded in the app — into a **navigable route**.
This would unlock “navigate this track” functionality without rewriting the core routing engine.
### **Key Components**
**Convert track to route**

- New menu option when tapping a line: Route to, Route from, Edit, Convert to Route.
- Works for both imported GPX tracks and locally recorded lines.

**Point density setting**

- User defines spacing between generated waypoints (e.g., 10 m, 250 m, 500 m, 1 km).
- Controls how closely the route hugs the underlying (OSM) road geometry.

**Routing proximity setting**

- Defines how strictly the routing engine snaps to those points (e.g., strict = 50 m, relaxed = 500 m).
- Can be adjusted globally (default) and per route (override).

**Per route adjustment**

- After conversion, user can tap the route anytime and change proximity settings.
- Keeps flexibility without forcing decisions at creation.

### **Benefits**
**Exact alignment:** Routes follow recorded or imported tracks closely.
**User control:** Adjustable density + proximity balance accuracy vs performance.
**Consistency:** Works for both imported GPX and recorded lines.
**Lightweight implementation:** Uses existing GPX <rte> standards and routing logic.
**High impact:** Adds “navigate this track” functionality with minimal code changes.
### **Example UI Flow**

- Record or import a line.
- Tap the line → menu shows Route to, Route from, Edit, Convert to Route.
- Dialog appears with two sliders:
    - Point spacing (100 m → 1 km).
    - Snap tolerance (strict → relaxed).
- Route is generated and stored with those parameters.
- Later, tap the route → Set Route Proximity to adjust snapping anytime.

<img width="967" height="935" alt="Image" src="https://github.com/user-attachments/assets/f1037720-0570-49a7-a64a-8f124303152f" />

**Comment by JohanJLugthart** [comment]
I would love it to connect the imported GPX track to the underlying map. OmsAnd has this option when you navigate. You select a GPX track to follow, and them it will travel on roads, giving normal road instructions. And it tries to follow the GPX track as good as possible.
I want to use my GPX tracks with a voice telling me how I need to go.

**Comment by JP7845** [comment]
Yes, I edited my comment to more specifically call out the OSM geometry. Thanks. In addition, I included an illustration to hopefully clarify and provide food for thought.

- User defines spacing between generated waypoints (e.g., 10 m, 250 m, 500 m, 1 km).
- Controls how closely the route hugs the underlying (OSM) road geometry.

<img width="1188" height="918" alt="Image" src="https://github.com/user-attachments/assets/4595a094-b810-4704-aa9a-c5574bb29b58" />

**Comment by Quantum-Future** [comment]
In your illustration, all generated points (converted track to toute), are on the GPX or recorded track, and don't align with the underlying nearest road. I guess this is because in this example, the proximity setting was "Relaxed".
If the user chooses "Strict", the generated points will be more closely aligned with the existing road geometry on OSM, right?

**Comment by JP7845** [comment]
In this example, I’m trying to describe what happens under the hood at the app level for developers. I think it’s accomplished in two steps: (1) **"Point density setting"** and (2) **"Routing proximity setting"**. I also included a side option to **flip the track/route**, so the user can start and finish wherever they prefer.

After importing a GPX/track file, the application may reduce or adjust the number of points based on what the user selects in the UI (for example, the “50m” setting shown below). From a processing standpoint, the goal is to control the number of points the app works with, and then use those points to generate a navigable route.

**Step 1: Point density setting**
The user sets the desired point density for the GPX/track. The app then processes the polyline and creates new points along the track at the interval specified by the user. This is important because GPX/track files can contain a large number of vertices, which can slow down performance. Once the GPX/track has been re-sampled this way, the app effectively creates a chain of “waypoints” along the route using the chosen density.

**Step 2: Routing proximity setting**
Next, the user sets how closely the route should follow the underlying OSM geometry. In practice, the app snaps the Step 1 waypoints to the nearest OSM vertices, but only within a user-defined proximity/tolerance. Tight proximity means the snapped points adhere more closely to the OSM network; looser proximity allows more flexibility when following OSM, depending on how far the OSM vertices are from the processed GPX vertices.

<img width="1584" height="1224" alt="Image" src="https://github.com/user-attachments/assets/bc87beb4-4073-4da8-ab93-1a1149818a18" />

**Comment by JP7845** [comment]
> In your illustration, all generated points (converted track to toute), are on the GPX or recorded track, and don't align with the underlying nearest road. I guess this is because in this example, the proximity setting was "Relaxed". If the user chooses "Strict", the generated points will be more closely aligned with the existing road geometry on OSM, right?

Does this answer your question? Sorry for the lengthy reply.

**Comment by Quantum-Future** [comment]
No problem for the lenghty reply, but it basically repeats (more in deep) what you said in first instance, wich I had already understood.
What I was asking is why the generated route (after steps 1 and 2) are not more closely aligned to the underlying existing road on the map.
I guess the answer could be: "Because the example doesn't use the maximum proximity setting". Am I correct?
Also I don't know the scale of the drawing and what is the supposed max distance between points recorded on GPX file and the real road.

**Comment by JP7845** [comment]
Thanks for the feedback:

- The dark green points are spaced at 50 m intervals as shown in the UI.
- The GNSS log will naturally fluctuate (accuracy, fixes, etc.), and the effective sample cadence is closer to ~15 m.

On why the route doesn’t follow the underlying OSM line:

- Since we snap/route using the 50 m (example ui) waypoint points, anything that falls between those positions (or outside that cadence) is effectively ignored.
- So, these points behave more like waypoints than an exact continuous line. When the system connects them, it looks like a route/line on screen, but it’s still driven by the waypoint spacing.

On “Routing proximity distance”:

- Setting it to 100 m allows the route to deviate in some cases, which is why a tighter setting produces behavior closer to the OSM geometry (in theory).
- There are still a lot of open questions about the routing engine. This approach was meant to avoid a full-year rewrite and deliver a faster improvement, but it’s definitely worth discussing and iterating.

I also put together an idea for non-OSM-based routes—unsure how that fits with the dev team, but I’m happy to share it if helpful.
Any other ideas or suggestions?

**Comment by Quantum-Future** [comment]
>     * The dark green points are spaced at 50 m intervals as shown in the UI.
I didn't pay attention to this info, sorry.  That's why I hadn't a distance reference on the map.

 

> 
> On “Routing proximity distance”:
> 
>     * Setting it to 100 m allows the route to deviate in some cases, which is why a tighter setting produces behavior closer to the OSM geometry (in theory).
> 
If the proximity setting is set at 100m, and having the reference that the distance between dark green points is 50m, now I understand why those dark green points are not closer to the real road. It's Ok.
If the user wants closer proximity, it may set the Proximity setting to 5 meters for example.

**Comment by JP7845** [comment]
> If the proximity setting is set at 100m, and having the reference that the distance between dark green points is 50m, now I understand why those dark green points are not closer to the real road. It's Ok.
> If the user wants closer proximity, it may set the Proximity setting to 5 meters for example.

We previously covered the data conversion process, but only partially. In the earlier graphic, the small magenta arrows and circles illustrate how each point is repositioned in Step 2 based on proximity. In this workflow, the underlying OSM data acts as the baseline for routing. The imported/collected data then snaps to the OSM network once proximity is applied.

Here, proximity works like a ring around every GPX/track point. For each point, the closest OSM geometry within that ring determines the point’s new location, and the point effectively ‘sticks’ to that snapped position. This behavior assumes a chain waypoint approach, where each waypoint’s updated position is determined through nearest-neighbor snapping to the OSM baseline.

<img width="1584" height="1224" alt="Image" src="https://github.com/user-attachments/assets/56cbbc8e-077b-4e4f-868f-740abc7e82d2" />

**Comment by JP7845** [comment]
Btw... good discussion. Thanks.

I hope the devs can use this info. This would be an amazing feature to add.

**Comment by Quantum-Future** [comment]
Great, now with the magnified image it's perfectly clear !

Agree: good discussion and hope this can be implemented by the devs. Not an easy task, but it's a much awaited feature by many users.

**Comment by TalkinBull** [comment]
Ride with GPS has this functionality built in if you want to see it working.  load a GPX file into the app on your phone and press the navigate button and it will convert the GPX into turn by turn directions.  Garmin divices do the same thing too.

**Comment by 0709wiwiwi** [comment]
[Track to TBT navigation track ](https://pitstopper.net/blog/match-your-ride-to-the-roads.html)
[The Navigation Hidden in Your GPX](https://pitstopper.net/blog/gpx-navigation-extensions.html)

_Pitstopper.net is NOT a navigation app. 
Import > Select ‘match to roads’ > Export as gpx > Import in Cruiser gps > Navigate > Go._
_Update: The function "match to roads" has been renamed into: "Track to Paths/Roads"_

**Comment by Quantum-Future** [comment]
Good info !
I hope the developers find inspiration in this info.
It's well explained how that software does it.

**Comment by biodranik** [comment]
The proposed settings dialog is way to complex. We don't plan to make yet another OSMand, so the solution should be simple and usable for your parents too: press a button, navigation works as expected. Without additional settings/actions.

**Comment by 0709wiwiwi** [comment]
> We don't plan

**Comment by Quantum-Future** [comment]
> [Track to TBT navigation track ](https://pitstopper.net/blog/match-your-ride-to-the-roads.html)
> [The Navigation Hidden in Your GPX](https://pitstopper.net/blog/gpx-navigation-extensions.html)
> 
> _Pitstopper.net is NOT a navigation app. Import > Select ‘math to roads’ > Export as gpx > Import in Cruiser gps > Navigate >

@biodranik , please look at this info.
Maybe it can be implemented in an almost automatic and simple way for the user.

**Comment by 0709wiwiwi** [comment]
- And a page which you can use to compare.
An original GPX (or FIT) track (eg from a GPS unit)
PitStopper's Track-to-Road converted version of your file
A 3rd GPX track (eg one converted elsewhere)
The tool will then point out the differences.
Download the PitStopper converted file and load in to ....
[Compare](https://pitstopper.net/compare-tracks.html)

- It's definitely worth keeping an eye on this as well.
[Changelog](https://pitstopper.net/staging/changelog.html)
By the way, “staging” is the Pitstopper beta version.

**Comment by JP7845** [comment]
@biodranik 

Yes, the proposed addition (see attached ui example) is two options as originally posted. Not knowing how the data is truncated maybe offering a settings dialog seemed like something to consider for longer tracks/GPX files. Just a thought. Thanks for the reply.

<img width="540" height="1170" alt="Image" src="https://github.com/user-attachments/assets/44c5219f-89ab-46cc-afd4-44c0757abfea" />

**Comment by sinnpi** [comment]
I've been experimenting around in implementing this feature, but don't feel like its quite ready to do a pull request on this official repo.

You can find it here: https://github.com/sinnpi/organicmaps-experimental/releases

It will install as a separate "Experimental Organic Maps" app. After installing you need to activate the feature in settings under the section "Experimental". Afterwards you can select some track, and it will show these new options when you tap "route to":

<img width="1440" height="818" alt="Image" src="https://github.com/user-attachments/assets/2278c588-57f3-41fd-a4f7-4729d6367284" />

Pretty self explanatory, but 'to this point' follows along the track to the selected point, the other two options to start or end of track respectively.

I'm not sure the algorithm is perfect yet, but it seems to correctly follow track wherever possible. But still a work in progress.

**Comment by biodranik** [comment]
@sinnpi is it the only UI change? How everything else looks/works? The good UI/UX is the hardest part here. I'm also experimenting with this task, and it's not quite easy and obvious.

Note that the core part should work for all platforms, in C++.

**Comment by sinnpi** [comment]
@biodranik sorry I forgot to say, this currently only runs on android. Problem is that I don't have any apple devices. All the track follow logic is in the backend though, so it should be fairly trivial to add the UI elements to the iOS version. All the track follow logic has of course been written in C++.

My fork has some other features that change the UI, but they're not relevant to this issue and can be isolated easily. But yes those are the only changes in UI, and they only show up when you actually select a track, not during normal navigation.

edit: I wanted to add, yes I realize UI is hard, and it is the main issue why I didn't do a pull request with the feature. I came to organicmaps because OsmAnd is so bloated (as you said), and I was looking for better alternatives. OrganicMaps seems like a major improvement, but you have to admit it is lacking some features. Having a simple to use interface that is still feature-rich is actually much harder than e.g. implementing a track follow algorithm. So I'm looking into what can be done to deliver features without making the interface complicated, but I feel like it could be done better than what I have so far.

**Comment by JP7845** [comment]
Hello @sinnpi , great job getting this idea in the app initially. I tested live yesterday and it worked. I used "follow track to this point". When I arrived it just terminated the routing with no dialog, which was fine. I really like the B/W option you added as well. Contrast works great while on a bike. I think this addition is very important for biking, hiking and motorcycles and being able to share your route and it be navigatable. Thanks for getting the ball rolling! @Quantum-Future FYI... give it a try.

**Comment by Quantum-Future** [comment]
Ok, I'll check it.

**Comment by 0709wiwiwi** [comment]
@sinnpi 
I’ve already tested the track navigation for a bit. (Foot-walk)
If you follow a track that runs along normal and accessible permitted OSM roads, it works just fine.

So good, in fact, that I quickly decided to test a slightly more challenging route.

- The first obstacle was following a track along private roads. At first, the tool refuses to follow the track. You can’t just enter private property simply like that. However, if you persevere for a moment, the tool will pick it up and continue navigating just fine. It then nicely guides you further until you leave the private roads again.

- The second obstacle was definitely much more challenging: following a drawn-in beeline or straight line where no OSM road exists. Here, the track was no longer followed. “Recalculating” kept appearing for a long time. Only once you than reach a drawn-in and permitted OSM road further along does the navigation finally pick up the track again and continue.

[Routing profile plus any path profile plus beeline.gpx.txt](https://github.com/user-attachments/files/32382538/Routing.profile.plus.any.path.profile.plus.beeline.gpx.txt)

**Comment by 0709wiwiwi** [comment]
I’ve already tested the track navigation for a bit. (Foot-walk)
If you follow a track that runs along normal, permitted OSM roads, it works just fine.

So good, in fact, that I quickly decided to test a slightly more challenging route.

- The first obstacle was following a track along private roads. At first, the tool refuses to follow the track. You can’t just enter private property like that. However, if you persevere for a moment, the tool will pick it up and continue navigating just fine. It then guides you until you leave the private roads again.

- The second obstacle was definitely much more difficult: following a drawn-in beeline or straight line where no OSM road exists. Here, the track was no longer followed. “Recalculating” kept appearing for a long time. Only once you reach a drawn-in and permitted OSM road further along does the navigation finally pick up the track again.

[Hike_any-path_beeline_hike.gpx.txt](https://github.com/user-attachments/files/32382917/Hike_any-path_beeline_hike.gpx.txt)

**Comment by 0709wiwiwi** [comment]
This [video](https://youtu.be/daU_vNZUrEU) shows the result as described. A virtual replay speed of 36 km/h was used solely to limit the record time.

Compare the result with an example of True track navigation (by integrated instructions into the trackpoints) as shown by the Crusier gps app.  [True track navigation](https://youtu.be/bfk_zglp9P4)

**Comment by sinnpi** [comment]
> * The second obstacle was definitely much more difficult: following a drawn-in beeline or straight line where no OSM road exists. Here, the track was no longer followed. “Recalculating” kept appearing for a long time. Only once you reach a drawn-in and permitted OSM road further along does the navigation finally pick up the track again.

I'm not sure what the solution for that would be. The purpose of the algorithm is to snap the track to the map data. I guess you could also turn the track into its own "virtual roads" that you follow, ignoring map data (just drawing map data, but not using it), but I feel like that kinda defeats the purpose of the app?

Anyways I did some 200km tour yesterday testing and the algorithm is definitely not perfect yet, it sometimes takes "shortcuts" and I don't yet fully understand why it is doing that. I'll look into it.

**Comment by 0709wiwiwi** [comment]
&sinnpi

The result is sure fine and will probably suffice for most people.

- Bmaps navigation.
It behaves very similarly. The app has an ‘I**gnore access rules**’ setting, which bypasses the normal (legal) route rules. As a result, Bmaps follows virtually every OSM road that has been mapped. Navigating on unmapped track sections is, of course, a bit ‘unusual’, and the Bmaps app does not follow such track sections either.

- Cruiser (Locus) 100% true track navigation.
See the above comparative video for this. It illustrates the difference. A well prepared GPX track import with attached TBT instructions is necessary. Additonally can contain announced via points or silent shaping points usefull to eventually auto(re)calculate the track or just a smaller tracksection.

**Comment by 0709wiwiwi** [comment]
@sinnpi 

The result is certainly fine and will probably be good enough for most people.
_The control menus are simple and clear._
With ‘Follow track to the end’, you simply follow any imported (recording) track.
_Preferably a track where at least the most important traffic rules are observed._
With ‘Follow track to the start’, this is not guaranteed.
_Make the necessary adjustments on the spot!_

Compare.
- Bmaps navigation.
It behaves very similarly. The app has an ‘**Ignore access rules**’ setting, which bypasses the normal (legal) route rules. As a result, Bmaps follows virtually every OSM road that has been mapped. Navigating on unmapped track sections is, of course, a bit ‘unusual’, and the Bmaps app does not follow such track sections either.

- Cruiser (Locus) 100% true track navigation.  Info: [The navigation hidden in your gpx](https://pitstopper.net/staging/blog/gpx-navigation-extensions.html)
See the above comparative video for this. It illustrates the difference. A well prepared GPX track import with attached TBT instructions is necessary. Additonally can contain announced via points or silent shaping points usefull to eventually auto(re)calculate the track or just a smaller tracksection.

**Comment by sinnpi** [comment]
@0709wiwiwi @JP7845 

There is a [new tag](https://github.com/sinnpi/organicmaps-experimental/releases/tag/experimental-2026.09.22-2) that is currently being built, APK should be available in a few minutes.

It fixes a bug where the algorithm accidentally took shortcuts (due to how the algorithm penalizes deviating a certain distance from the track) that should be better now. But needs more testing.

It also adds a new option under "Routing options" (the little gear wheel that shows up when you made a route) that allows "Ignore access restrictions". This only shows up if you selected a track. Careful, it might route you through private property, or to a locked gate, or similar. Use at your own discretion.

Edit: Build failed. I'll look into it later

Edit2: [New tag](https://github.com/sinnpi/organicmaps-experimental/releases/tag/experimental-2026.09.22-2) is up and built. Give it a go when you have time.

**Comment by JP7845** [comment]
@sinnpi Great work! I will test further. Thank you! 

Just curious, I haven't tried this but will this allow routing along a trail? What if that trail is a GPX file that has been downloaded? Or does the routing logic stick to the nearest "OSM road" only?

**Comment by sinnpi** [comment]
@JP7845 If the trail/path is in OSM (where I live OSM is the best navigation app for hiking) and you choose pedestrian, it should work without a problem. But I haven't tried it yet.

**Comment by hoizan** [comment]
Kudos! It works well but the UX parts needs to improve to make this feature more discoverable. Currently, users get this new "navigating saved tracks" menu when clicking the 'Route to' button only after selected a point along a saved/imported track, or viewing a selected track on map. One idea is to add this new option menu in to the track main menu items list, so that users have a new option to navigate besides edit, hide, export and delete a track.

**Comment by sinnpi** [comment]
@hoizan That is a reasonable proposal and I've implemented it in the [latest tag](https://github.com/sinnpi/organicmaps-experimental/releases/tag/experimental-2026.09.24-1).

<img width="1440" height="3088" alt="Image" src="https://github.com/user-attachments/assets/2ee036df-ebb9-48df-8fe1-3be9cec3c34d" />

**Comment by hoizan** [comment]
Thank you for your consideration and implementation 🙏
